darts/explainability/tft_explainer.py

```diff
@@ -0,0 +1,249 @@
+from typing import Dict, List, Literal, Optional
+
+import matplotlib.pyplot as plt
+import pandas as pd
+from torch import Tensor
+
+from darts import TimeSeries
+from darts.explainability.explainability import (
+    ExplainabilityResult,
+    ForecastingModelExplainer,
+)
+from darts.models import TFTModel
+
+
+class TFTExplainer(ForecastingModelExplainer):
+    def __init__(
+        self,
+        model: TFTModel,
+    ):
+        """
+        Explainer class for the TFT model.
+
+        Parameters
+        ----------
+        model
+            The fitted TFT model to be explained.
+        """
+        super().__init__(model)
+
+        if not model._fit_called:
+            raise ValueError("The model needs to be trained before explaining it.")
+
+        self._model = model
+
+    @property
+    def encoder_importance(self):
+        return self._get_importance(
+            weight=self._model.model._encoder_sparse_weights,
+            names=self._model.model.encoder_variables,
+        )
+
+    @property
+    def decoder_importance(self):
+        return self._get_importance(
+            weight=self._model.model._decoder_sparse_weights,
+            names=self._model.model.decoder_variables,
+        )
+
+    def get_variable_selection_weight(self, plot=False) -> Dict[str, pd.DataFrame]:
+        """Returns the variable selection weight of the TFT model.
+
+        Parameters
+        ----------
+        plot
+            Whether to plot the variable selection weight.
+
+        Returns
+        -------
+        TimeSeries
+            The variable selection weight.
+
+        """
+
+        if plot:
+            # plot the encoder and decoder weights
+            self._plot_cov_selection(
+                self.encoder_importance,
+                title="Encoder variable importance",
+            )
+            self._plot_cov_selection(
+                self.decoder_importance,
+                title="Decoder variable importance",
+            )
+
+        return {
+            "encoder_importance": self.encoder_importance,
+            "decoder_importance": self.decoder_importance,
+        }
+
+    def explain(self, **kwargs) -> ExplainabilityResult:
```

Comment:
How about taking some of the `predict()` parameters explicitly? At least `series`, `past_covariates`, `future_covariates` and `n` would make sense IMO. It will produce more comprehensible API documentation.

Reply:
I am not sure if that is relevant here at all. 

I do not understand why predict has to be called to get the proper attention heads of the time series. The learned autoregressive connections should depend on how `predict` is called. But if `predict` is not called at all the `attention_heads` saved in `self._model.model._attn_out_weights` do not have the right format. I assume they are still in a state of training and the ` predict()` call changes that. 

If that is the case I would rather remove the `**kwargs` completely from the `explain` method here and call predict once with `self._model.model.output_chunk_length` to get the correct attention heads.

Reply:
Yes I agree with you we need to call `predict()` here. However `predict()` takes a lot more arguments than just `n`. It takes `series` (the series to predict), as well as covariates arguments and other arguments: see [the API doc](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.tft_model.html#darts.models.forecasting.tft_model.TFTModel.predict). 
I think we should probably change the signature of `explain()` to something like
```python
def explain(self, series, past_covariates, future_covariates, **kwargs) -> ExplainabilityResult
```
This in the docstring you can list `series`, `past_covariates` and `future_covariates`, and explain that those are passed down to `predict()`. You can also say that `n` will always be set to `output_chunk_length` (unless I'm wrong I think that's always what's needed), and that `**kwargs` can contain extra arguments for the predict method and link to the API documentation of `TFTModel.predict()`.
I hope it makes sense.

Reply:
I think calling `predict()` is just a technicality to get to the correct attention weights. I don't think the way we call `predict` matters at all for the result, its just important that it was called (for whatever reason). 
If I understand it correctly the attention weights are learned during training and are not impacted by the data used in the `predict` call.

They don't have a similar logic behind them like shapley values but are learned during the training and are a fixed part of the trained model. 

Maybe I am wrong, but if I am right I would rather remove all parameter passed to `explain()` and have the `predict()` call happen without the user needing to know about it at all.